Support importing simpleperf trace files from Android Studio

nisargjhaveri opened the pull request:
Fix #5203.

Attempt to add importer for simpleperf trace files. These can be generated using the `simpleperf report-sample` command or using Profiler UI in Android Studio.

Here is the file format details: https://cs.android.com/android/platform/superproject/main/+/main:system/extras/simpleperf/cmd_report_sample.proto

Added a couple of dependencies to work with protobuf. These will also be useful for more importers in future, I believe pprof also uses protobufs (in discussion at #4462).

This is one attempt. I'm fairly new to the codebase and Flow. Feel free to suggest any changes as required.

mstange commented:
Can you share an example output profile?

This importer would be nice to have. As an alternative you can also use `samply import perf.data` with a samply built from the main branch of https://github.com/mstange/samply .

julienw commented on src/profile-logic/import/simpleperf.js:
```diff
@@ -0,0 +1,644 @@
+  }
+
+  findOrAddResource(file: report.IFile): IndexIntoResourceTable {
+    if (!this.resourcesMap.has(file.id)) {
```
If you do something such as :
```js
let resourceIndex = this.resourcesMap.get(file.id);
if (!resourceIndex) {
  ...
  resourceIndex = this.resourceTable.length++;
  this.resourcesMap.set(file.id, resourceIndex);   
}
return resourceIndex;
```
then you don't need your `$NotNull` utility, with the added bonus that the hashmap is requested only once.

Same comment for all the code in this file that applies.

julienw commented on src/profile-logic/import/simpleperf.js:
```diff
@@ -0,0 +1,644 @@
+class FirefoxResourceTable {
+  strings: UniqueStringArray;
+
+  resourceTable: ResourceTable = {
```
It would be better to use the functions `getEmptyXXX` from src/profile-logic/data-structures.js.

This comment applies to all the other tables as well.

julienw commented on src/profile-logic/import/simpleperf.js:
```diff
@@ -0,0 +1,644 @@
+    this.name = thread.threadName ?? '';
+  }
+
+  toJson(): SerializableThread {
```
Any reason you decided to use a `SerializableThread` (and `SerializedProfile` below) instead of a normal `Thread` and (`Profile`)?

nisargjhaveri replied:
I migrated this code from somewhere else where I had to use SerializableProfile. Changed now.

julienw commented on src/profile-logic/import/simpleperf.js:
```diff
@@ -0,0 +1,644 @@
+    }
+
+    this.sampleTable.stack.push(prefixStackId);
+    $NotNull(this.sampleTable.time).push(toMilliseconds(sample.time ?? 0));
```
$NotNull is probably not needed if you use a Thread instead of a SerializedThread

julienw commented on src/profile-logic/import/simpleperf.js:
```diff
@@ -0,0 +1,644 @@
+      symbolicationNotSupported: true,
+      markerSchema: [],
+
+      // platform: "Android",
+      // device?: string,
```
Just wondering if these comments were supposed to be uncommented?

nisargjhaveri replied:
Sorry for this. I was playing around. We don't really have "device" information and the `platform` param here seems to make no difference in UI, or at lease I don't know how it is used. Planning to remove, let me know if it should be set to something more sensible.

julienw replied:
This is how we handle the "platform" information: https://github.com/firefox-devtools/profiler/blob/ae149d7439f6a15e2088a8c20f58474415af1f35/src/profile-logic/profile-metainfo.js#L58-L65

Very imperfect :-) but you can probably leverage this here.

nisargjhaveri replied:
This helped. Added both `toolkit` and `platform` to meta.

julienw commented on src/profile-logic/import/simpleperf.js:
```diff
@@ -0,0 +1,644 @@
+    const thread = this.threadMap.get(sample.threadId);
+
+    if (!thread) {
+      // logger.warn(`Thread not found for sample: ${sample.threadId}`);
```
could be good to keep this log with `console.warn`.

julienw commented on src/profile-logic/import/simpleperf.js:
```diff
@@ -0,0 +1,644 @@
+  bufferView: DataView;
+  bufferOffset: number = 0;
+
+  static verifyMagic(traceBuffer: ArrayBuffer): boolean {
+    return (
+      new TextDecoder('utf8').decode(
+        traceBuffer.slice(0, this.magic.length)
+      ) === this.magic
+    );
+  }
```
if we want to be able to lazy load this file, it could be good to move this to a separate file. My suggestion would be to have src/utils/magic.js with just this simple tool. This file could also export various magic constants (currently just this one).

Then from the `process-profile` file we could do:
```js
import { verifyMagic, SIMPLEPERF as SIMPLEPERF_MAGIC } from "firefox-profiler/utils/magic";

...

if (verifyMagic(SIMPLEPERF_MAGIC)) {
 ...
}
```

nisargjhaveri replied:
Updated.

julienw commented on src/profile-logic/process-profile.js:
```diff
@@ -1882,6 +1886,8 @@ export async function unserializeProfileOfArbitraryFormat(
       if (isArtTraceFormat(arrayBuffer)) {
         arbitraryFormat = convertArtTraceProfile(arrayBuffer);
+      } else if (isSimpleperfTraceFormat(arrayBuffer)) {
+        arbitraryFormat = convertSimpleperfTraceProfile(arrayBuffer);
```
it would be good to use dynamic imports so that `import/simpleperf` would be lazyloaded. This way we wouldn't have protobufjs and long.js in the main bundle (they're pretty big!).

julienw reviewed (changes requested):
Thanks, this looks pretty good! Impressive work!

I suggested a few things, tell me what you think!
Also it would be good to have a few tests, at least one import test. See unit/profile-conversion.test.js for existing tests.

nisargjhaveri commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "private": true,
   "engines": {
-    "node": ">= 18.16 < 19"
+    "node": ">= 18.16"
```
Revert this, unintentional change

nisargjhaveri commented:
@mstange Thanks for pointing out samply, seems interesting. I'll have a look! Though this only helps when you have control over the perf.data file which might not be available when exporting the profile from Android Studio.

@julienw Thanks a lot for the review. I have addressed most of the suggestions. Please have a look?

julienw reviewed (approved):
Thanks, this looks good to me!
I rebased and updated some part of the code that were needed after the rebase and I'm gonna land this soon!
Thanks for your patience

julienw commented:
Would you want to add some limited documentation around https://profiler.firefox.com/docs/#/./guide-android-profiling ?

nisargjhaveri commented:
> Would you want to add some limited documentation around https://profiler.firefox.com/docs/#/./guide-android-profiling ?

Thanks @julienw for updating and merging. Created #5309 to update the documentation.
